[WIP] Documents ETL [WIP] (#20)
CVTS/cvts: doc/source/etl.rst

**************
Data ingestion
**************

The raw data provided by HANEL comes in a format that is inadequate for most, if not all, analytics tasks performed
with the data. For this reason,  processing the data into a format appropriate for our use is the first step in the
generation of any analytical outputs.

This process is commonly known as `Extract, Transform & Load <https://en.wikipedia.org/wiki/Extract,_transform,_load>`_
and can (should) be designed with the analytics use cases in mind. However, there is absolutely no need to run this
process more than once on the raw data

.. _hanel_data:

HANEL's raw data
================

The data received from HANEL has three important characteristics that make it inadequate for use for analytics in its
original format:

> 1. It is sorted by timestamp rather than by vehicle, while most of the processing of the data is done byy vehicle

Reviewer: ```suggestion
1. It is sorted by timestamp rather than by vehicle, while most of the processing of the data is done by vehicle
```

    n3sdB/si5JemE8XMifymLA==,1585766618,0.0,105.2251966667,21.1300333333,272.0,Xe chưa phân loại

Data transfer
-------------

There is still no agreed-upon procedure for data transfer from HANEL to the FPT servers, and all transfers
done to this point have been in an ad-hoc fashion. Further work for designing this process will be necessary
should recurring data transfers be needed.

.. _backup:

Backup
======

Since the process of obtaining data from HANEL was expensive and time-consuming, it was reasonable to
safeguard the data by backing up the raw data received in a format closest to the original so we
could begin the processing from scratch at any point in time.

Backing up the data in its original CSV format would require an unreasonable amount of disk space, so
we did a little investigation on different formats we could use for such task, for which the results
are shown below.

All these tests were done with Pandas, as it provides a simple interface for loading the original CSV
and saving in all of these formats and the times are for converting a single 10 million row CSV.

.. image:: images/compression.png
    :width: 960
    :alt: Compression performance

> All all these formats, Zip, GZ and BZ2 are dominated by other alternatives, and therefore were discarded

Reviewer: ```suggestion
Three formats, Zip, GZ and BZ2 are dominated by other alternatives, and therefore were discarded
```

threads, it was possible to backup an entire month worth of data in less than 4h.

.. _etl:

Extract, Load & Transform
=========================

Due to the volume of data and its organization, and to allow for the leveraging of as many threads as available in
our infrastructure the data ingestion was divided into two phases:

1. Separation of the original data into one file per vehicle
2. Consolidation of daily files into one file per vehicle per month

During development, we verified that loading the Parquet files generated during the backup process is substantially
faster than loading the original CSV files, and therefore we load the data from the backup rather than from the
original CSV files.

.. _datalake_database:

Reference database
------------------

Although we have opted to leave all the GPS data on disks as compressed files, as it will be discussed in the
following sessions, we rely on the existence of a reference database to maintain a record of the correspondence between
the original vehicle identifiers and their integer counterparts on our system, as well as the vehicle type of each
of the vehicles for which data has been loaded.

> We also leverage the database to store summary statistics of for the data, particularly on the table *vehicle_days*,

Reviewer: ```suggestion
We also leverage the database to store summary statistics of the data, particularly in the table *vehicle_days*,
```

where we have one record for each day a vehicle has been active. For each one of these records we also include
the number of GPS pings, the first and last active instants and the bounding box of the vehicle's GPS trace as a
Polygon.

The point of this table is to provide a more powerful index of the data, as it allows for the filtering of the raw
data to vehicles that operate during a certain period and region, as well as discard those which operate during a
period too short or with too few GPS records to be deemed useful.

.. _daily_processing:

Separating traces per vehicle
-----------------------------
The first process in the ingestion of the raw data provided by HANEL is to separate all data in separate datafiles
for each vehicle available in the dataset. Since a large portion of the work is data compression

The data format chosen for the vehicle specific datasets was GZIP, as Pandas allows one to stream the data into
pre-existing files, enabling us to process each one of (or a set of) the original datasets at a time, cumulatively

> streaming the data into the original data files.

Reviewer: ```suggestion
streaming the data into the output data files.
```
